h/services/oauth_validator.py

```diff
@@ -21,20 +22,45 @@ class OAuthValidatorService(RequestValidator):
     This implements the ``oauthlib.oauth2.RequestValidator`` interface.
     """
 
-    def __init__(self, session):
+    def __init__(self, session, user_svc):
         self.session = session
+        self.user_svc = user_svc
 
         self._cached_find_client = lru_cache_in_transaction(self.session)(self._find_client)
+        self._cached_find_refresh_token = lru_cache_in_transaction(self.session)(self._find_refresh_token)
 
     def authenticate_client_id(self, client_id, request, *args, **kwargs):
         """Authenticates a client_id, returns True if the client_id exists."""
         client = self.find_client(client_id)
         request.client = client
         return (client is not None)
 
+    def client_authentication_required(self, request, *args, **kwargs):
+        """
+        Require client authentication for confidential clients.
```

Comment:
I think this would be clearer as something like "Determine if client authentication is required for an access token request".